modules/fedimint-wallet-client/src/backup/recovery_history_tracker.rs

```diff
@@ -0,0 +1,128 @@
+    // most of which will be thrown away, ignore script pubkeys from before this `session_idx`
+    decoy_session_threshold: u64,
+}
+
+impl ConsensusPegInTweakIdxesUsedTracker {
+    pub(crate) fn next_unused_tweak_idx(&self) -> TweakIdx {
+        self.next_unused_tweak_idx
+    }
+
+    pub(crate) fn new(
+        previous_next_unused_idx: TweakIdx,
+        start_session_idx: u64,
+        current_session_count: u64,
+
+        data: &WalletClientModuleData,
+    ) -> Self {
+        debug_assert!(start_session_idx <= current_session_count);
+
+        let mut s = Self {
+            pending_pubkey_scripts: BTreeMap::new(),
+            next_pending_tweak_idx: previous_next_unused_idx,
+            decoys: VecDeque::new(),
+            decoy_session_threshold: current_session_count
+                .saturating_sub((current_session_count.saturating_sub(current_session_count)) / 20),
+            next_unused_tweak_idx: previous_next_unused_idx,
+        };
+
+        s.init(data);
+
+        s
+    }
+    fn generate_next_pending_tweak_idx(&mut self, data: &WalletClientModuleData) {
+        let (script, _address, _tweak_key, _operation_id) =
+            data.derive_peg_in_script(self.next_pending_tweak_idx);
+
+        self.pending_pubkey_scripts
+            .insert(script, self.next_pending_tweak_idx);
+        self.next_pending_tweak_idx = self.next_pending_tweak_idx.next();
+    }
+
+    fn init(&mut self, data: &WalletClientModuleData) {
+        for _ in 0..RECOVER_MAX_GAP {
+            self.generate_next_pending_tweak_idx(data);
+        }
+        debug_assert_eq!(self.pending_pubkey_scripts.len(), RECOVER_MAX_GAP as usize);
+    }
+
+    fn remove_and_replenish_by_tweakidx(
+        &mut self,
+        data: &WalletClientModuleData,
+        tweak_idx: TweakIdx,
+    ) {
+        if let Some(script) = self
+            .pending_pubkey_scripts
+            .iter()
+            .find(|(_k, v)| *v == &tweak_idx)
+            .map(|(k, _v)| k.clone())
+        {
+            self.pending_pubkey_scripts.remove(&script);
+            self.generate_next_pending_tweak_idx(data);
+        }
+        debug_assert_eq!(self.pending_pubkey_scripts.len(), RECOVER_MAX_GAP as usize);
+    }
+
+    pub(crate) fn handle_script(
+        &mut self,
+        data: &WalletClientModuleData,
+        script: &bitcoin::ScriptBuf,
+        session_idx: u64,
+    ) {
+        if let Some(tweak_idx) = self.pending_pubkey_scripts.remove(script) {
+            debug!(target: LOG_CLIENT_MODULE_WALLET, %session_idx, ?tweak_idx, "Found previously used tweak_idx in federation history");
+            self.generate_next_pending_tweak_idx(data);
+
+            let mut to_remove = self.next_unused_tweak_idx;
+            while to_remove <= tweak_idx {
+                self.remove_and_replenish_by_tweakidx(data, tweak_idx);
+                to_remove = to_remove.next();
+            }
```

Comment:
Oh, good catch.

W.r.t using in order - since the request was to improve recovery privacy, my initial thought was that all I need is to find last claimed idx, and then query the blockchain from there to find pegged-in, but yet unclaimed addresses. Since we need to expose ability to recheck deposits to particular address in the apps, I thought that is good enough.

But indeed I can maintain a collection of all indexes that were already claimed instead, and then just skip querying blockchain for deposits on them. I don't really ever need to remove anything from the list of addresses I'm looking for - even if there's >10k of them, it shouldn't matter much anyway. And querying long ago unused addresses is not much of a privacy issue, as they will probably not get used in the future anyway.